library/core/src/panicking.rs

```diff
@@ -109,6 +111,27 @@ fn panic_no_unwind() -> ! {
     unsafe { panic_impl(&pi) }
 }
 
+/// Aborts the process, but with a properly displayed panic message.
+#[cold]
+pub(crate) fn panic_abort<'a>(message: Option<&'a fmt::Arguments<'a>>) -> ! {
```

Comment:
This function must itself guard against unwind, in case the panic handler unwinds (`panic_no_unwind` above is automatically nounwind as a special case in codegen attribute calculation code in the compiler).

You can mark this function as `extern "C"`, mark with `#[rustc_allocator_nounwind]` (despite its name it's not allcoator-specific at all) or make it a lang item and yet another special case in the compiler.

Reply:
Interesting! Will do (I guess I'll go with the attribute). But now I'm curious: is that mostly as an optimization (better codegen for callers) or is it semantically required? (also cc @CAD97)

Reply:
Well, in this PR it's not semantically required, since this code will only be hit in double-panic situation which will abort anyway even for a normal panic.

But if the caller assumes that `panic_abort` can't unwind then it'll semantically required. I think it's highly likely that this function will have other use sites in the future, so it's better to ensure that it's guarded ahead of time.

Reply:
I'm *fairly* sure that there's no way that this function could unwind anyway, as that would require one of the four called functions to unwind:

- `intrinsics::abort`
- `Location::caller`
- `PanicInfo::internal_constructor`
- `#[lang = "panic_impl"] panic_impl`

and I'm *fairly* sure that all of those already have a soundness-critical requirement to not unwind, or at the very least (e.g.  in the case of `panic_impl`) unwinding invariably causes a reentrant unwind and thus abort.

...but adding `#[rustc_allocator_nounwind]` is still a good idea, since it forces the function to not unwind and serves as a bit of self-documentation and safeguarding against mistakes.

Reply:
`panic_impl` can easily unwind via FFI-unwind with a custom `#[panic_handler]` or panic hook.